tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/AsyncFallbackTest.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.fault.tolerance.tck;
+
+import org.eclipse.microprofile.fault.tolerance.tck.asynchronous.CompletableFutureHelper;
+import org.eclipse.microprofile.fault.tolerance.tck.asynchronous.fallback.AsyncFallbackClient;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.annotations.Test;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+
+/**
+ * Test the combination of the @Asynchronous and @Fallback annotations.
+ */
+public class AsyncFallbackTest extends Arquillian {
+    @Inject
+    private AsyncFallbackClient client;
+
+    @Deployment
+    public static WebArchive deploy() {
+        JavaArchive testJar = ShrinkWrap
+                .create(JavaArchive.class, "ftAsyncFallback.jar")
+                .addClasses(AsyncFallbackClient.class, CompletableFutureHelper.class)
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .as(JavaArchive.class);
+
+        WebArchive war = ShrinkWrap
+                .create(WebArchive.class, "ftAsyncFallback.war")
+                .addAsLibrary(testJar);
+        return war;
+    }
+
+    @Test
+    public void testAsyncFallbackSuccess() throws ExecutionException, InterruptedException {
+        assertThat("Future-returning method that returns successfully should not fallback",
+            client.service1().get(), equalTo("Success"));
+    }
+
+    @Test
+    public void testAsyncFallbackMethodThrows() throws IOException, ExecutionException, InterruptedException {
+        assertThat("Future-returning method that throws an exception should fallback",
+            client.service2().get(), equalTo("Fallback"));
+    }
+
+    @Test
+    public void testAsyncFallbackFutureCompletesExceptionally() throws InterruptedException {
+        try {
+            client.service3().get();
+        }
+        catch (ExecutionException expected) {
+            assertThat("Future-returning method that returns failing future should not fallback",
+                expected.getCause(), instanceOf(IOException.class));
+        }
+    }
+
+    @Test
+    public void testAsyncCSFallbackSuccess() throws ExecutionException, InterruptedException {
+        assertThat("CompletionStage-returning method that returns successfully should not fallback",
+            client.serviceCS1().toCompletableFuture().get(), equalTo("Success"));
```

Comment:
`CompletionStage.toCompletableFuture()` is technically an optional method. Could you use the static method on  `CompletableFutureHelper` instead?

Reply:
I've seen the helper method and I thought it's superfluous, because I believed that all `CompletionStage`s support it. I read the javadoc again and you're right, it says "A CompletionStage implementation that does not choose to interoperate with others may throw UnsupportedOperationException." I'll fix this.

Reply:
Fixed.